drivers: input: Input subsystem

MichaelOwen2022 opened the pull request:
Input subsystem support api for input devices (like keyboard, touscreen, etc).

rosterloh commented on drivers/key/key_gpio.c:
```diff
@@ -0,0 +1,343 @@
+	const struct key_info_dt_spec *info = &config->info[idx];
+	struct key_gpio_driver *driver = &data->driver[idx];
+
+	bool pressed = (gpio_pin_get(gpio->port, gpio->pin) ^ gpio->dt_flags);
```
This driver works great for me but this line causes pressed to always be true and so each event only ever fires once. This is on an stm32l4

mbolivar-nordic commented on include/zephyr/drivers/key.h:
```diff
@@ -0,0 +1,123 @@
+ */
+
+/**
+ * @file
```
No new APIs without documentation in Sphinx. You need to update doc/develop/api/overview.rst for this API, and add a page documenting this new API in .rst just like the other APIs mentioned in overview.rst.

mbolivar-nordic commented on dts/bindings/gpio/gpio-keys.yaml:
```diff
@@ -15,3 +15,15 @@ child-binding:
           required: true
           type: string
           description: Human readable string describing the device (used as device_get_binding() argument)
+       code:
+          required: false
+          type: int
+       ms_debounce:
+          required: false
+          type: int
+       ms_long:
+          required: false
+          type: int
+       ms_hold:
+          required: false
+          type: int
```
cc @petejohanson since I'm guessing this is keyboard related

mbolivar-nordic commented on include/zephyr/drivers/key.h:
```diff
@@ -0,0 +1,123 @@
+}
+
+/**
+ * @brief Remove a key instance.
```
From what? Where? This documentation isn't enough to understand what's going on.

mbolivar-nordic commented on include/zephyr/drivers/key.h:
```diff
@@ -0,0 +1,123 @@
+ */
+
+/**
+ * @brief Setup a key instance.
```
What does this do? Please fix this documentation.

mbolivar-nordic commented on include/zephyr/dt-bindings/key/key.h:
```diff
@@ -0,0 +1,98 @@
+#define KEY_KP0                 82
+#define KEY_KPDOT               83
+
+// ... ...
```
Please remove

mbolivar-nordic commented on samples/basic/gpio_key/src/main.c:
```diff
@@ -0,0 +1,86 @@
+
+	key = DEVICE_DT_GET_ONE(gpio_keys);
+
+	if (key == NULL) {
+		printk("key device is not exist\n");
+		return;
```
This can't happen

mbolivar-nordic reviewed (dismissed):
In addition to the problems here, you need to fix your commit logs.

Please watch https://www.youtube.com/watch?v=mnVNLM7ynMQ or at least read the slides for an overview of the contribution process: https://static.sched.com/hosted_files/zephyr2022/a3/HOWTO_Get_Your_Zephyr_Patches_Merged_ZDS_2022.pdf

MichaelOwen2022 commented on drivers/key/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_KEY_GPIO key_gpio.c)
+
+zephyr_library_sources_ifdef(CONFIG_USERSPACE key_handlers.c)
```
This is in reference to some other drivers (like led, kscan).
I don't understand how to fix it?

MichaelOwen2022 commented on drivers/key/key_handlers.c:
```diff
@@ -0,0 +1,23 @@
+#include <zephyr/drivers/key.h>
+#include <zephyr/syscall_handler.h>
+
+static inline int z_vrfy_key_setup(const struct device *dev, key_callback_t callback_isr)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_KEY(dev, setup));
+	Z_OOPS(Z_SYSCALL_VERIFY_MSG(callback_isr == 0, "callback cannot be set from user mode"));
+	return z_impl_key_setup((const struct device *)dev, callback_isr);
+}
+#include <syscalls/key_setup_mrsh.c>
```
This is in reference to some other drivers (like kscan, sensor).
How do you recommend reporting keyboard events?

dcpleung replied:
Setting callbacks should not be exposed to userspace, and should run in kernel mode. So please remove syscall for the setup function.

dcpleung reviewed (changes requested):
Can I suggest to rename this to `keyboard` instead of just `key`? This would make it more clear that these are keyboard related stuff.

str4t0m commented:
Tagging @petejohanson as he might be interested and have some feedback.

petejohanson commented:
I think if this is intended just for simple "send a key" uses, this will work, but for anything as complex as https://github.com/zmkfirmware/zmk/ this abstraction would not work for us.

* ZMK settled on using the existing kscan driver API, and implemented both matrix and direct wire versions. There seems a lot of overlap (potentially) between this new API and kscan.
* By using KSCAN + GPIO abstraction that exists, our existing matrix driver, for instance, works fine with IO expanders, shift registers, etc. This, in theory, might also be able to do that, but something to be aware of.
* On the above point this seems to mix a few concerns:
  * Assignment of "key values" to different hardware.
  * Debouncing
  * Hardware details
* Debounce can actually be complicated, and should ideally not be duplicated. Eager debounce can be used to get key presses sent earlier, and debounce the release instead, etc.

Just a few thoughts, I'll think more and add more when I can.

MichaelOwen2022 commented:
@petejohanson @dcpleung
Whether this commit has a chance to be merged?
If this commit doesn't have a chance to merge, I will close this pull request.
If I implement a new input module instand of kscan, userspace use poll mothed to get input events, like linux input subsystem. Is that can be merged?

dcpleung commented:
My issue is simply that setting up the callback should not be done from user thread. It can be done in kernel mode.

MichaelOwen2022 commented:
> My issue is simply that setting up the callback should not be done from user thread. It can be done in kernel mode.

I mean, If I solve the callback program, is there a chance to merge? this pull request is highly similar to kscan driver. 

@dcpleung

dcpleung commented:
If it is similar to kscan as you said, would it be possible to roll this into kscan instead?

carlescufi commented:
cc @desowin

carlescufi commented:
@petejohanson  could you please review this? Thanks!

pdunaj commented on drivers/input/input_handlers.c:
```diff
@@ -0,0 +1,74 @@
+}
+#include <syscalls/input_config_set_mrsh.c>
+
+static inline int z_vrfy_input_event_read(const struct device *dev, struct input_event *event)
```
async api would be very useful
blocking api forces you to use a dedicated thread to handle read.

I suggest creating a pollable sync object that can be used similarly as in linux poll.
Here is the example from NCS how it can be done https://github.com/nrfconnect/sdk-nrf/blob/main/samples/nfc/tag_reader/src/main.c#L703

pdunaj commented on drivers/input/input_internal.c:
```diff
@@ -0,0 +1,185 @@
+	return retval;
+}
+
+int input_event(struct input_dev *dev, uint16_t type, uint16_t code, int32_t value)
```
creating an event system only for input may be a resource waste.
I would suggest integrating this with system wide event manager. We have the code ready for that (handling input events for both usb and ble) in NCS.

pdunaj commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,408 @@
+		}
+		break;
+
+	case KEY_STATE_LONG_PRESSED:
```
I suggest separating interaction events from button handler - make things ready for future expansion.
Here is code from NCS (again based on event manager) https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/caf/modules/click_detector.c

pdunaj commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,408 @@
+		driver->count_cycle++;
+	}
+
+	if (driver->pressed != pressed) {
```
besides debounce keyboard also needs deghosting https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/caf/modules/buttons.c#L261

pdunaj reviewed (commented):
The code should allow creating keyboard sample. I have added some suggestions that should prepare it for integration. Hope that helps.

pdunaj commented:
One more comment - I think this would look better in subsystems (input subsystem) then drivers.

pdunaj commented:
One more comment - I think this would look better in subsystems (input subsystem) then drivers.

pdunaj commented:
If somebody is looking for nice input subsystem build on top of modular architecture (for simplified application development) I recommend looking at https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/caf/index.html

Small adv but maybe it will help somebody.

MichaelOwen2022 commented:
@dcpleung could you please review this again? Thanks!

desowin commented on include/zephyr/dt-bindings/input/input_key_code.h:
```diff
@@ -0,0 +1,753 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_INPUT_INPUT_KEY_CODE_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_INPUT_INPUT_KEY_CODE_H_
+
+/*
```
This is simply copied from [linux-event-codes.h](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/input-event-codes.h?h=v5.19#n64) without any attribution. Note that the linux header is licensed under `GPL-2.0-only WITH Linux-syscall-note`

desowin commented on drivers/input/Kconfig:
```diff
@@ -0,0 +1,32 @@
+	bool "Enable input isr lock"
+	help
+	  Use isr_lock() and isr_unlock() to control ring buffer exclusive access.
+	  This is use to report input event in isr context.
```
The description is quite confusing. "This is required to be able to report input event in isr context."  might be better.

desowin reviewed (changes requested):
I don't like the use of GPLv2 code without any attribution.

jfischer-no reviewed (changes requested):
As already mentioned by @desowin, file [input_key_code.h](https://github.com/zephyrproject-rtos/zephyr/blob/d1486c8ede8bb7b5e084837f40afdb6f48655288/include/zephyr/dt-bindings/input/input_key_code.h) is a copy of [linux-event-codes.h](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/input-event-codes.h?h=v5.19#n64). The author also claims the copyright on it, which is very doubtful. This PR should be closed.

MichaelOwen2022 commented:
I'm just don't understand how to use lisense rules.
@desowin @jfischer-no

MichaelOwen2022 commented:
@desowin  could you please review this again? Thanks!

desowin commented:
> @desowin could you please review this again? Thanks!

This cannot be merged due to licensing. Regarding licensing the easiest way is to contribute original code that you have created yourself and hence are free to use any license you wish (it is recommended to choose Apache-2.0 which is used by Zephyr).

If your work was created in part by others, then it is important that every author has signed off the [Developer Certificate of Origin](https://developercertificate.org/) and you have to make sure the license is compatible. [The FSF has never considered the Apache License to be compatible with GPL version 2](https://www.apache.org/licenses/GPL-compatibility.html). Note that while GPLv3 is compatible with Apache-2.0, the whole work would have to be licensed under GPLv3 and this is not suitable for Zephyr.

MichaelOwen2022 commented:
> > @desowin could you please review this again? Thanks!
> 
> This cannot be merged due to licensing. Regarding licensing the easiest way is to contribute original code that you have created yourself and hence are free to use any license you wish (it is recommended to choose Apache-2.0 which is used by Zephyr).
> 
> If your work was created in part by others, then it is important that every author has signed off the [Developer Certificate of Origin](https://developercertificate.org/) and you have to make sure the license is compatible. [The FSF has never considered the Apache License to be compatible with GPL version 2](https://www.apache.org/licenses/GPL-compatibility.html). Note that while GPLv3 is compatible with Apache-2.0, the whole work would have to be licensed under GPLv3 and this is not suitable for Zephyr.

All codes is written by me, except input_key_code.h.
What should I do with input_key_code.h? Modify Lisense directly or rewrite this header file?

MichaelOwen2022 commented:
> > @desowin could you please review this again? Thanks!
> 
> This cannot be merged due to licensing. Regarding licensing the easiest way is to contribute original code that you have created yourself and hence are free to use any license you wish (it is recommended to choose Apache-2.0 which is used by Zephyr).
> 
> If your work was created in part by others, then it is important that every author has signed off the [Developer Certificate of Origin](https://developercertificate.org/) and you have to make sure the license is compatible. [The FSF has never considered the Apache License to be compatible with GPL version 2](https://www.apache.org/licenses/GPL-compatibility.html). Note that while GPLv3 is compatible with Apache-2.0, the whole work would have to be licensed under GPLv3 and this is not suitable for Zephyr.



> > @desowin could you please review this again? Thanks!
> 
> This cannot be merged due to licensing. Regarding licensing the easiest way is to contribute original code that you have created yourself and hence are free to use any license you wish (it is recommended to choose Apache-2.0 which is used by Zephyr).
> 
> If your work was created in part by others, then it is important that every author has signed off the [Developer Certificate of Origin](https://developercertificate.org/) and you have to make sure the license is compatible. [The FSF has never considered the Apache License to be compatible with GPL version 2](https://www.apache.org/licenses/GPL-compatibility.html). Note that while GPLv3 is compatible with Apache-2.0, the whole work would have to be licensed under GPLv3 and this is not suitable for Zephyr.

I will rewrite input_key_code.h，and check the licenses of all files, thank you for your help.

carlescufi commented:
> I will rewrite input_key_code.h，and check the licenses of all files, thank you for your help.

This should work. You cannot take any GPL-licensed code and pull it into the Zephyr tree, but rewriting it and using the Linux code for inspiration is acceptable.

MichaelOwen2022 commented:
@desowin @carlescufi could you please review this again? Thanks!

Commit:
drivers: input: add input subsystem.

Input subsystem support public API for input device.

Signed-off-by: Michael Owen <tangchunhui@coros.com>

Commit:
drivers: input: keyboard: add gpio keyboard driver.

This driver scan gpio keyboard status,
and report key event to input subsystem.

Signed-off-by: Michael Owen <tangchunhui@coros.com>

Commit:
samples: keyboard: add gpio keyboard sample.

This sample use input api to get gpio keyboard event.

Signed-off-by: Michael Owen <tangchunhui@coros.com>

MichaelOwen2022 commented:
* Run compliance checks on patch series (PR)
* Scan code for licenses
How can I run above two checks automatically or locally? 

What's wrong with code not being reviewed?
@carlescufi

desowin commented on doc/develop/api/overview.rst:
```diff
@@ -376,3 +376,8 @@ current :ref:`stability level <api_lifecycle>`.
      - 2.0
+
+   * - :ref:`input_api`
+     - Stable
```
Newly introduced subsystem cannot be classified as stable. See [API Lifecycle](https://docs.zephyrproject.org/3.2.0/develop/api/api_lifecycle.html#stable) for more information.

desowin commented on drivers/input/Kconfig:
```diff
@@ -0,0 +1,36 @@
+config ENABLE_INPUT_ISR_LOCK
+	bool "Input event interrupt lock"
+	help
+	  Use isr_lock() and isr_unlock() to control ring buffer exclusive access.
```
I think this should be converted to a spinlock. See [Legacy irq_lock() emulation](https://docs.zephyrproject.org/3.2.0/kernel/services/smp/smp.html#legacy-irq-lock-emulation) for more information.

desowin commented:
> What's wrong with code not being reviewed?

The problem is not necessarily in code, but atleast in my case it is the amount of pending work. I am still getting familiar with Zephyr and for example, in my latest review I have mentioned the API Lifecycle and irq_lock() emulation issues that I had no idea about earlier. Zephyr is pretty big project and it is not easy to pick up everything straight away. However frustrating it might be, the review is necessary evil as otherwise projects of such size and complexity will quickly turn into unmaintainable monsters.

MichaelOwen2022 commented:
> > What's wrong with code not being reviewed?
> 
> The problem is not necessarily in code, but atleast in my case it is the amount of pending work. I am still getting familiar with Zephyr and for example, in my latest review I have mentioned the API Lifecycle and irq_lock() emulation issues that I had no idea about earlier. Zephyr is pretty big project and it is not easy to pick up everything straight away. However frustrating it might be, the review is necessary evil as otherwise projects of such size and complexity will quickly turn into unmaintainable monsters.

I mean after the code is recommitted, it takes a long time for someone to review the new code.
Do I need to @ someone to review the code, or what else do I need to do ?

I modified the code according to your suggestion, can you reivew it again ?

In addition, I am considering whether it is necessary to report the event in the interrupt ？Do you have any suggestions？

carlescufi commented:
@MichaelOwen2022 sorry, this flew under my radar. Could you please rebase?
I've asked for a re-review from the other reviewers.

MichaelOwen2022 commented:
> @MichaelOwen2022 sorry, this flew under my radar. Could you please rebase? I've asked for a re-review from the other reviewers.

I've rebase already.

tbursztyka commented on drivers/input/input_internal.c:
```diff
@@ -0,0 +1,214 @@
+
+#include "input_internal.h"
+
+int input_internal_setup(struct input_dev *dev)
```
From one input device stand-point: are you going to call this more than once at runtime?

If not, then all of this can be integrated into a macro with relevant static intialization for each objects (rb, mutex, sem...)

tbursztyka commented on include/zephyr/drivers/input.h:
```diff
@@ -0,0 +1,307 @@
+ */
+__syscall int input_setup(const struct device *dev);
+
+static inline int z_impl_input_setup(const struct device *dev)
```
Why can't this be done at device initialization? (see previous comment)

tbursztyka commented on include/zephyr/drivers/input.h:
```diff
@@ -0,0 +1,307 @@
+ */
+__syscall int input_event_read(const struct device *dev, struct input_event *event);
+
+static inline int z_impl_input_event_read(const struct device *dev, struct input_event *event)
```
as noted already, adding asynchronous input API would be a must here.

tbursztyka commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,421 @@
+	KEY_EVENT_NONE = 0xff,
+};
+
+struct kbd_gpio_driver {
```
weird naming, the whole code in keyboard_gpio.c is the driver. 
Actually I don't see the point of not including of these attributes into kbd_gpio_data

tbursztyka commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,421 @@
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_attr_get(input, type, attr);
```
retval seems superfluous, just return from the called input_* function

tbursztyka commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,421 @@
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_attr_set(input, type, attr);
```
same here

tbursztyka commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,421 @@
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_event_read(input, event);
```
same here

tbursztyka commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,421 @@
+	struct input_dev *input = data->input;
+	int retval = 0;
+
+	retval = input_internal_event_write(input, event);
```
same here

tbursztyka commented on include/zephyr/drivers/input.h:
```diff
@@ -0,0 +1,307 @@
+ */
+__syscall int input_release(const struct device *dev);
+
+static inline int z_impl_input_release(const struct device *dev)
```
what would be the use case for this function?

Like the setup above. It seems to bring concepts (pluggin in/out a keyboard? pm perhaps?) to the user level that should not be its role to manage. Which is not the right way.

tbursztyka reviewed (changes requested):
Input is a way too generic name, a keyboard being one of the many input devices (and we already have kscan for this).
Hard to see how this could be used for a mouse, or else...

MichaelOwen2022 commented:
> Input is a way too generic name, a keyboard being one of the many input devices (and we already have kscan for this). Hard to see how this could be used for a mouse, or else...

With all due respect, I don't think you're good enough to understand this input subsysem framework.

petejohanson commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,421 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT keyboard_gpio
```
I still don't know why this isn't done as a kscan driver, and this input system doesn't have a mapping to consume a kscan driver and map to input system events.

MichaelOwen2022 replied:
I don't think you understand enough of this input subsystem framework either.

petejohanson commented on drivers/input/keyboard/keyboard_gpio.c:
```diff
@@ -0,0 +1,421 @@
+			continue;
+		}
+
+		err = gpio_pin_interrupt_configure_dt(gpio, GPIO_INT_EDGE_TO_ACTIVE);
```
Not all gpio pins support interrupts. This will have limitations and only work on certain hardware configurations.

Also, using edge to active interrupt won't allow deep sleep/wake from deep sleep on nRF52.

petejohanson commented on include/zephyr/drivers/input.h:
```diff
@@ -0,0 +1,307 @@
+enum keyboard_value {
+	KEY_RELEASE,
+	KEY_PRESSED,
+	KEY_LONG_PRESSED,
```
Why are these added. As soon as you have long press/release, why not support double tap? Triple? 

If the goal is a "lowest common denominator" API good enough for basic input to programs, this is probably fine.

I think maybe this is my disconnect, I come from developing featureful input devices. 

If the goal of this API is to focus on the *consumers* of input data, then we should clearly document that, since the current system doesn't seem to be flexible/powerful enough to handle the other.

petejohanson commented on include/zephyr/dt-bindings/input/input.h:
```diff
@@ -0,0 +1,94 @@
+
+/* keyboard code */
+
+#define KEY_CODE_RESERVED   0
```
The current code set may be fine for a basic sample program, but definitely won't allow a complex system that covers the full range of possible keyboard codes, other possible layouts/languages, etc.

petejohanson commented on include/zephyr/dt-bindings/input/input.h:
```diff
@@ -0,0 +1,94 @@
+#define KEY_CODE_Y          89
+#define KEY_CODE_Z          90
+
+#define KEY_CODE_KP0        96
```
There's far more keypad codes than these.

MichaelOwen2022 commented:
It's been a few months since this commit and the code review that followed was disappointing. I can't speak highly of the entire zephyr team and have no interest in committing anything on this project.

MichaelOwen2022 commented:
It's no fun to waste time with a bunch of elementary school students。
